### PR TITLE
[WIP] constant blend of latest-podman + latest-buildah

### PR DESCRIPTION
**Pull request by edsantiago**
Followup to Cabal discussion 2022-04-07, re: letting us vendor in buildah on a moment's notice.

Instead of cron'ing this, let's see if this works: keep this PR open perpetually, with daily updates to get latest podman and buildah.

See individual commit messages for details.

### CODE REVIEW

**Comment by edsantiago** [comment]
The first commit is the important one. This one must be hand-crafted by a human, there is no possible way to automate it. This is the one that, in case of emergency vendor, can be grabbed into a real PR.

The second commit is a throwaway, ignore it. That's just automated fetch-the-latest-buildah.

**Comment by edsantiago** [comment]
bud test failed, [remote only](https://storage.googleapis.com/cirrus-ci-6707778565701632-fcae48/artifacts/containers/podman/5927716103913472/html/bud-remote-fedora-35-root-host.log.html#t--00037), because the new `--identity-label=false` option is not getting passed through. Probably an easy fix, but not this late in my day.

This is still a proof-of-concept PR, and my feeling so far is positive: it has quickly identified multiple problems in the buildah merge that are easy to solve iteratively. A nightly cron job, or a non-gating step in buildah CI, would be utterly useless because neither one would have the crucial first-commit. Without that, there's no way to make it past the first or second or third failure.

**Comment by rhatdan** [comment]
I just pushed a fix for this.

**Comment by Luap99** [comment]
How is this PR regularly updated?

**Comment by edsantiago** [comment]
By a human, presumably me. There is actually no other way: this is not possible to solve via cron or PRs.

I am working on a script to automate as much of it as possible.

**Comment by edsantiago** [comment]
Followup, because I don't want anyone to think this is abandonware: it's not. I'm running my script at least twice a day, one of those times in the evening. One of the nice features I added is:
```
+---> Podman has bumped, but Buildah is unchanged. There's probably not much point to testing this.
```
So I'm choosing to save CI cycles by not re-pushing today. Should anyone need to re-vendor buildah tomorrow or this weekend (when I'm unavailable), just cherry-pick the first commit from this PR. It's still valid.

**Comment by edsantiago** [comment]
Yippee, script has caught a bug! https://github.com/containers/buildah/pull/3919 breaks when vendoring into podman:
```
+---> Running 'make' to confirm that podman builds cleanly...
CGO_ENABLED=1 GOOS=linux GOARCH=amd64 go build \
        -mod=vendor  \
        -ldflags '-X github.com/containers/podman/v4/libpod/define.gitCommit=3bf05d3113df29addd392032797f4fdf5dae56b7 -X github.com/containers/podman/v4/libpod/define.buildInfo=1650539506 -X github.com/containers/podman/v4/libpod/config._installPrefix=/usr/local -X github.com/containers/podman/v4/libpod/config._etcDir=/usr/local/etc -X github.com/containers/common/pkg/config.additionalHelperBinariesDir= ' \
        -tags "   selinux systemd  exclude_graphdriver_devicemapper seccomp" \
        -o bin/podman ./cmd/podman
# github.com/containers/podman/v4/libpod
libpod/events.go:18:3: cannot use r.config.Engine.EventsLogFileMaxSize (type "github.com/containers/common/pkg/config".eventsLogMaxSize) as type uint64 in field value
make: *** [Makefile:321: bin/podman] Error 2
buildah-vendor-treadmill-x: 'make' failed with new buildah. Cannot continue.
```
I could just force-push right now and let y'all see the breakage in CI, or I could just report it like I am right here. Or I could file an issue in containers/common since that's where the offending code seems to be?

Anyhow, @rhatdan et al, PTAL because this is not going to vendor into podman.

**Comment by edsantiago** [comment]
What the heck, I pushed anyway in case someone wants to pull & play.

**Comment by Luap99** [comment]
@edsantiago this is expected, we have to update c/common in podman https://github.com/containers/podman/pull/13945

**Comment by vrothberg** [comment]
> What the heck, I pushed anyway in case someone wants to pull & play.

Apologies, @edsantiago!

**Comment by edsantiago** [comment]
No need for apologies, this is a useful test of the script. It also helps me understand, because I thought the script's vendor-new-buildah would also pull in new c-c. It does, `v0.47.5-0.20220420095823-d822f53650b2`, but your PR pulls in an even newer one, and I assume everything will work again as soon as your PR merges.

**Comment by Luap99** [comment]
@edsantiago It pulls in the new one but this version is incompatible with podman that is why it fails at compile time. We have to tweak the code in podman before it will work.

**Comment by edsantiago** [comment]
Today's push is failing, seems to be related to the new `etchosts` code in buildah. I'm going to assume that this is Someone Else's Problem, likely one that they already know about, and it will be fixed in a day or two.

**Comment by Luap99** [comment]
I have no idea why it would fail, however I am sure it will be fixed once #13918 is merged

**Comment by edsantiago** [comment]
New `--output` option needs to be implemented in podman.

**Comment by edsantiago** [comment]
@containers/podman-maintainers heads-up:
```
* bin/podman grew by 63360 bytes; max allowed is 51200.
```
I have temporarily bypassed in order to run tests. I am sending this as an early notice for your awareness. The culprit is https://github.com/containers/buildah/pull/3949 . (We don't have bloat checks in buildah, but it's safe to assume that this vendor update caused similar growth in the buildah binary).

**Comment by rhatdan** [comment]
Do you know where the bloat came from? Containers/storage?

**Comment by edsantiago** [comment]
See my comment in the recent containers-storage vendor PR. Yes, 45K was that PR, the rest must've been buildah. There is no way to know beyond that.

**Comment by edsantiago** [comment]
[Remote failed](https://storage.googleapis.com/cirrus-ci-6707778565701632-fcae48/artifacts/containers/podman/5105045082472448/html/int-remote-fedora-35-root-host.log.html#t--Remote-build--containerignore-filtering-embedded-directory-13535---1); could be a flake, but it's not one I've seen before. Could someone TAL?

**Comment by edsantiago** [comment]
Latest buildah fails:
```
xref-helpmsgs-manpages: 'podman build --help' lists '--env', which is not in docs/source/markdown/podman-build.1.md
xref-helpmsgs-manpages: 'podman build --help' lists '--os-feature', which is not in docs/source/markdown/podman-build.1.md
xref-helpmsgs-manpages: 'podman build --help' lists '--os-version', which is not in docs/source/markdown/podman-build.1.md
xref-helpmsgs-manpages: 'podman image build --help' lists '--env', which is not in docs/source/markdown/podman-build.1.md
xref-helpmsgs-manpages: 'podman image build --help' lists '--os-feature', which is not in docs/source/markdown/podman-build.1.md
xref-helpmsgs-manpages: 'podman image build --help' lists '--os-version', which is not in docs/source/markdown/podman-build.1.md
```
Caught by the treadmill script, and normally this is something I'd easily fix, but I'm on PTO. If you're gonna be vendoring in buildah this week, I've got some of the work done for you here (use the treadmill script) but this new `--env` one, you're on your own, sorry.

**Comment by rhatdan** [comment]
No problem I will take care of it.

**Comment by edsantiago** [comment]
Heads-up that this morning's treadmill run failed in a way I've never seen before:
```
+---> Vendoring in buildah...
go mod tidy
warning: ignoring symlink /home/esm/src/atomic/2018-02.podman/libpod/contrib/systemd/user
warning: ignoring symlink /home/esm/src/atomic/2018-02.podman/libpod/docs/build/man/links
warning: ignoring symlink /home/esm/src/atomic/2018-02.podman/libpod/src
go: downloading github.com/containers/common v0.48.1-0.20220608111710-dbecabbe82c9
go: downloading github.com/containers/buildah v1.26.1-0.20220609034125-a818b5527f02
go: finding module for package github.com/docker/libnetwork/ipamutils
go: downloading github.com/docker/libnetwork v0.5.6
go: found github.com/docker/libnetwork/ipamutils in github.com/docker/libnetwork v0.5.6
warning: ignoring symlink /home/esm/src/atomic/2018-02.podman/libpod/contrib/systemd/user
warning: ignoring symlink /home/esm/src/atomic/2018-02.podman/libpod/docs/build/man/links
warning: ignoring symlink /home/esm/src/atomic/2018-02.podman/libpod/src
go: finding module for package github.com/Sirupsen/logrus
go: downloading github.com/Sirupsen/logrus v1.8.1
go: found github.com/Sirupsen/logrus in github.com/Sirupsen/logrus v1.8.1
go: github.com/containers/podman/v4/pkg/specgenutil imports
        github.com/docker/docker/opts imports
        github.com/docker/libnetwork/ipamutils imports
        github.com/docker/libnetwork/osl imports
        github.com/Sirupsen/logrus: github.com/Sirupsen/logrus@v1.8.1: parsing go.mod:
        module declares its path as: github.com/sirupsen/logrus
                but was required as: github.com/Sirupsen/logrus
make: *** [Makefile:255: vendor] Error 1
```
It worked fine yesterday, so this is something that changed between yesterday and right now. Thought y'all might want the early alert.

**Comment by edsantiago** [comment]
@containers/podman-maintainers heads up, treadmill is reporting that the next buildah vendor will trigger the bloat block.

>* bin/podman grew by 56712 bytes; max allowed is 51200.

**Comment by edsantiago** [comment]
Yo people, bloat is getting heavier. It was 56712 just two days ago, now it's:

>* bin/podman grew by 77216 bytes; max allowed is 51200.

**Comment by edsantiago** [comment]
bloat is now:

>bin/podman grew by 83048 bytes; max allowed is 51200.

**Comment by flouthoc** [comment]
I'll remove it from `podman`.

**Comment by flouthoc** [comment]
> I'll remove it from `podman`.

PR here: https://github.com/containers/podman/pull/14907

**Comment by edsantiago** [comment]
Good news, the `manifest list` error is fixed. Bad news, `bind to hostIP` failed in [int f36 root](https://api.cirrus-ci.com/v1/artifact/task/5247280457449472/html/int-podman-fedora-36-root-host.log.html#t--podman-run-network-bind-to-HostIP--1) with what looks like yet another networking flake:
```
# podman [options] run --network slirp4netns:outbound_addr=10.88.0.1,allow_host_loopback=true -dt quay.io/libpod/alpine:latest nc -w 2 10.0.2.2 38291
         Ncat: Version 7.92 ( https://nmap.org/ncat )
         Ncat: Listening on :::38291
         Ncat: Listening on 0.0.0.0:38291
         608a4c7ced331482bc14c35a3c5dbd3faae732d667d42bdbc470aa8ac04acf0a
         Ncat: Connection from 10.128.0.11.
         Ncat: Connection from 10.128.0.11:57953.
         [AfterEach] Podman run networking
           /var/tmp/go/src/github.com[/containers/podman/test/e2e/run_networking_test.go:37](https://github.com/containers/podman/blob/178b444300a74c83104938e4d8a91358d08aa789/test/e2e/run_networking_test.go#L37)
         # podman [options] pod rm -fa -t 0
         # podman [options] rm -fa -t 0
         608a4c7ced331482bc14c35a3c5dbd3faae732d667d42bdbc470aa8ac04acf0a
         
    ...
           Expected
               <string>: Ncat: Version 7.92 ( https://nmap.org/ncat ) Ncat: Listening on :::38291 Ncat: Listening on 0.0.0.0:38291 Ncat: Connection from 10.128.0.11. Ncat: Connection from 10.128.0.11:57953.
           to contain substring
               <string>: 10.88.0.1
```

**Comment by edsantiago** [comment]
Oh wow
>* bin/podman grew by 1113040 bytes; max allowed is 51200.

Also, macintosh docs are failing:
```
time="2022-07-13T13:43:08-07:00" level=error msg="stat /usr/share/containers/storage.conf: no such file or directory"
docs: no doc file nor link docs/build/man/podman-.1 for 'podman '
```

**Comment by edsantiago** [comment]
The containers/storage vendor changed a few error messages. 

What is the plan for vendoring containers/storage into podman? Will it happen as part of a buildah vendoring, or standalone (i.e., #14892)? I'm asking because I don't want to duplicate effort by fixing tests that are being fixed elsewhere.

**Comment by rhatdan** [comment]
I think the storage and buildah will have to be done together.  Since storage change required change to Buildah tests.

**Comment by rhatdan** [comment]
If you can get them to work, I will push for a merge. Sadly I am wrapped up in meedings all day today.

**Comment by edsantiago** [comment]
@containers/podman-maintainers @rhatdan this is as ready as I can get it. Only remaining problem is the Macintosh one, which is not something I can fix.

Next step for vendoring buildah: https://github.com/containers/podman/wiki/Buildah-Vendor-Treadmill#for-vendoring-buildah-start-here (and remember that the `--pick` step isn't well tested, so please send me bug reports or fixes)

**Comment by ashley-cui** [comment]
Looking at the Mac issue now..

**Comment by ashley-cui** [comment]
Okay, I can reproduce it and confirmed that the failure doesn't exist on main. It's a docs build error, which is strange since no docs have been changed in this PR. Digging a little further, but this is strange..

**Comment by edsantiago** [comment]
If it helps, I think this is a hint:
>docs: no doc file nor link docs/build/man/podman-.1 for 'podman '
This suggests that `podman_all_commands` isn't generating a useful list; which in turn suggests that `podman_commands` isn't working, which suggests that `podman help` is not spitting out something that looks like "Available Commands ...." Can you try running `podman help` and `podman help ps` and things like that?

**Comment by ashley-cui** [comment]
Hmm, I think another hint is that it's looking for `podman-.1` and `'podman '` (note the trailing space and trailing dash). But I need to dig on why its broken here and not in main... Podman help does spit out Available Commands correctly.

**Comment by edsantiago** [comment]
That's the same thing: the symptom you note (spaces dash) is because `podman_all_commands` isn't working. But I have no way to find out why.

**Comment by rhatdan** [comment]
Do we still substiture podman-remote.1.md for podman.1.md on MAC and Windows?

**Comment by ashley-cui** [comment]
Found out what's gone wrong: Podman entirely is broken with this PR, all podman commands fail with `ERRO[0000] stat /usr/share/containers/storage.conf: no such file or directory`, so the script isn't finding any available commands to go through. This was introduced in https://github.com/containers/buildah/commit/7c0f2253f2a8746a93ece173c3df19e88dd76c72. I haven't dug deep enough to know too much further where the error is coming from but I'd assume it's from common or storage. I can dig further if you'd like, but it'll probably be a few hours of vendor dancing to see which commit broke podman. Let me know if you want me to do that :)

**Comment by edsantiago** [comment]
A very very quick `git diff` suggests https://github.com/containers/storage/pull/1279 because it does all sorts of `err` returns. And a very quick look at the repo suggests that we might want to pull in https://github.com/containers/storage/pull/1287 to see if that fixes it.

**Comment by edsantiago** [comment]
Made it past the OSX test this time. Assuming the rest of the tests pass, I'd say buildah vendor is Go. Bloat is now:

>* bin/podman grew by 1,132,936 bytes; max allowed is 51200.

**Comment by edsantiago** [comment]
(No point in restarting the APIv2 error, it's simply the deadlock bug)

**Comment by Luap99** [comment]
@edsantiago heads up there is a regression somewhere in c/image or c/storage, https://github.com/containers/podman/pull/15695

**Comment by edsantiago** [comment]
@Luap99 checking in: were you able to get the treadmill script to work, after that patch? Any suggestions for improvement? As I mentioned, I've got a big cleanup PR in the works.

**Comment by Luap99** [comment]
> Any suggestions for improvement? As I mentioned, I've got a big cleanup PR in the works.

There were some small problem, first I had to use a GITHUB_TOKEN without it I always got a `401 unauthorized` from github query back (I guess not s problem with your script and github doesn't like my IP instead)

Also at the last step I assume you want the user to edit your commit message right?
```
This commit was automatically cherry-picked
by %s v%s
from the buildah vendor treadmill PR, #%s

/vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
> The git commit message from that PR is below. Please review it,
> edit as necessary, then remove this comment block.
\^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

However below this message there was nothing so I had to copy it manually in.

**Comment by edsantiago** [comment]
>However below this message there was nothing so I had to copy it manually in.

Oops. I changed one part of the code without making corresponding fixes elsewhere. Fixed.

The `GITHUB_TOKEN` thing is good to know. I will improve that error message. Thank you!

**Comment by edsantiago** [comment]
failing with latest buildah commits:
```
=== RUN   TestShellCompletionFunctions
    shell_completion_test.go:56: podman.test build --skip-unused-stages is a bool flag but has a shell completion function set.
        You have to remove this shell completion function.
    shell_completion_test.go:56: podman.test image build --skip-unused-stages is a bool flag but has a shell completion function set.
        You have to remove this shell completion function.
```
https://github.com/containers/podman/pull/13808/checks?check_run_id=8415396905

@Luap99 PTAL when you have a moment. I'd call this low priority right now.

**Comment by edsantiago** [comment]
Tested locally by removing the offending completion line in my `vendor/buildah`. Seems to work. Submitted https://github.com/containers/buildah/pull/4256

**Comment by Luap99** [comment]
> Tested locally by removing the offending completion line in my `vendor/buildah`. Seems to work. Submitted [containers/buildah#4256](https://github.com/containers/buildah/pull/4256)

That is correct, however I should port the podman test to buildah so podman will not be blocked in the future by this.

**Comment by edsantiago** [comment]
That would be much appreciated!

BTW, to whomever cares, the new `--skip-unused-stages` (buildah 4249) is not working in podman. Tests fail. ISTR that something changed in the last months, such that buildah options should magically Just Work in podman; if so, it's not working in this particular case.

Just a heads-up, no need for anyone to do anything about it this week.

**Comment by Luap99** [comment]
https://github.com/containers/buildah/pull/4267 should improve the buildah test

**Comment by edsantiago** [comment]
The new `--dns*` conflict warnings (buildah 4250) are also failing tests. Looks like the conflict checks are not happening in podman.

**Comment by edsantiago** [comment]
Lots of new `skip`s. Three new options need to be wired in. Sorry I can't do that.

**Comment by edsantiago** [comment]
This morning's treadmill failed with hundreds of lines of "blah blah blah go 1.16 would select something else". So, no push today. If someone is going to vendor buildah into podman today, be sure to remove the two `skip_if_remote` lines that @flouthoc fixed yesterday.

**Comment by edsantiago** [comment]
lint failures due to c/storage. Heads-up to whoever does the next c/storage revendor into podman.

**Comment by edsantiago** [comment]
OK, treadmill is back. Ignore the remote-rootless hang, I've been trying (off and on) for weeks to track that one down. Rootless bud tests only run in this treadmill PR, not in regular podman nor gating, so that failure does not affect anything in the real world.

**Comment by edsantiago** [comment]
@containers/podman-maintainers heads-up, new vendoring is broken:
```
[+1077s] not ok 338 podman-system-service --log-level=trace should be able to hijack
...
# podman --url tcp://127.0.0.1:5628 run --name c-4ybQQEoIDK quay.io/libpod/testimage:20221018 echo o-Y1TT0H7d7R
Error: Post "http://d/v4.6.0/libpod/containers/create": dial tcp 127.0.0.1:5628: connect: connection refused
```
Consistent hard failure. New vendored modules are buildah, c/common, c/image, c/storage, logrus, hcsshim, containerd, crypto, oauth2.

Also, bloat.

**Comment by flouthoc** [comment]
I think it happened after vendor of `c/common, c/image, c/storage` , cause I created a PR before the vendoring commit and it passes https://github.com/containers/podman/pull/18579

**Comment by edsantiago** [comment]
It could be. These are the [go.mod diffs](https://github.com/containers/podman/compare/8e253ef4be0c36128853f161b86117ccd95e23bd..9a69b30ceb424beb2f8c4499768868de4a72a2bc#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6) between the last working treadmill run, and last night's failing one.

**Comment by flouthoc** [comment]
@edsantiago I think failing test is not using any of the build component, it could be either `c/common` or `golang.org/x/oauth2`. I will try it out and confirm which this is causin the issue.

**Comment by flouthoc** [comment]
@edsantiago I think https://github.com/containers/buildah/pull/4806 should fix the issue.

**Comment by edsantiago** [comment]
@containers/podman-maintainers, @containers/buildah-maintainers, heads-up: latest containers/storage seems to be breaking buildah. Failures filed in #18631, see there for list of symptoms. Although most prevalent in debian, failures also seen in f38.

**Comment by edsantiago** [comment]
CI [still failing](https://api.cirrus-ci.com/v1/artifact/task/5968970201694208/html/sys-podman-debian-12-rootless-host-boltdb.log.html), even with the new containers/storage.

**Comment by flouthoc** [comment]
@edsantiago How can i know buildah version inside the `podman` CI, if buildah version got updated on the VM and it points before `c/storage` `9ef15e4d49b76a9af9f548e9415153009faa54d8` and after `e3b18ab721f02f801d93806ac789e958cbe6a050` it will keep choosing wrong `storage` driver.

**Comment by flouthoc** [comment]
Yes same tests are failing , maybe this is related https://github.com/containers/podman/pull/18510#discussion_r1189812306

**Comment by nalind** [comment]
The newer versions of podman, in an environment where neither an older version of podman nor an older version of buildah have been run before, is still going to select overlay, and after that, the older version of buildah is still going to select vfs.  In a test, buildah stores images and containers using vfs, podman looks for them using overlay, and so the test fails.

The storage change avoids breaking the case where an older version of buildah or podman was run _before_ the newer version of podman, altering the storage library's behavior so that the newer version of podman doesn't decide to use overlay despite indications on disk that vfs is already being used.  I expect older buildah, like almost-newer podman, is likewise ignoring any tracks that newer podman leaves because that part of the storage library's initialization logic had this problem for a long time.

If we decide to have newer podman, in an environment where an older version of podman or buildah has _not_ been previously run, still select vfs instead of overlay, then I think that sounds equivalent to reverting containers/storage#1571, and I'm reluctant to do that because it provides a tangible benefit for cases where there's no configuration file.

**Comment by edsantiago** [comment]
I apologize for taking so long to respond. I've read your comment twice a day every day and am still having trouble understanding it. My best guess is that we have a situation, right now (May/early June 2023) in which a user can:
1. install and run podman; then
2. later on, install and run buildah; then
3. some things will break, because the currently-available podman and buildah do not play well with each other.

Is that correct? Have I missed something? If that is correct, what steps are being taken to correct this situation? (And if I've misunderstood, I'm sorry, but could someone please ELI5?)

**Comment by flouthoc** [comment]
> * install and run podman; then
> 
> * later on, install and run buildah; then
> 
> * some things will break, because the currently-available podman and buildah do not play well with each other.

That sounds correct to me. The older buildah version ( which does not contains necessary patches from `c/storage` will keep causing this issue ) .  The `c/storage` change was a breaking change for the particular use-case which @nalind mentioned in the first para here: https://github.com/containers/podman/pull/13808#issuecomment-1568815229 and `debian` VM in `podman` CI contains old `buildah` version so it'll keep breaking.

I am not sure what is the best way to fix this but may I suggest back-porting necessary patches of `c/storage` to older `buildah` and `podman` version.  @nalind @edsantiago WDYT ?

**Comment by flouthoc** [comment]
> Is the only problem here only the mismatched storage driver? If so why not just create /etc/containers/storage.conf with overlayfs as driver.

That would solve the issue in CI afaics, but issue mentioned here https://github.com/containers/podman/pull/13808#issuecomment-1568815229 will persists for users who used created setup without `storage.conf` just like its breaking in the CI now.

@Luap99 but yes I agree that we should unblock CI if everybody is cool with it.

**Comment by edsantiago** [comment]
Heads-up, something broke in `secret replace`, vendoring into podman is broken right now.

**Comment by flouthoc** [comment]
> Heads-up, something broke in `secret replace`, vendoring into podman is broken right now.

@edsantiago Is it still broken, does this needs investigation ?

**Comment by edsantiago** [comment]
Yesterday's run went fine, so I think all is fixed. Thanks @flouthoc.

**Comment by cevich** [comment]
@edsantiago is this still used?  I noticed a task `rootless_buildah_bud_test_task` in `.cirrus.yml` on main referencing this issue.  However it runs `only_if: $CIRRUS_CRON == 'treadmill'` and that cirrus-cron job is no longer active.  Just wondering in case that task can be stripped out of main to simplify the YAML a bit.

**Comment by edsantiago** [comment]
@containers/buildah-maintainers PTAL. Something in latest buildah is breaking podman int tests. As best I can tell, the common factor is that buildah is now re-pulling images that are already present:
```
old (good):
$ podman build ...
STEP 1 2 3 etc
```
```
new (bad):
$ podman build ...
Trying to pull .....  <<<<<<<<<< SHOULD NOT HAPPEN
```
(I may be wrong in my analysis. Maybe the common factor is something else I didn't notice)

**Comment by nalind** [comment]
> @containers/buildah-maintainers PTAL. Something in latest buildah is breaking podman int tests. As best I can tell, the common factor is that buildah is now re-pulling images that are already present:

It looks like the image caching save/load sequence is changing the compression used for the image's layer, and thus the manifest, and thus the manifest's digest, and when `build` checks if the image in the registry has a different digest when compared to the one it has on disk, it sees that it is and pulls it again.

To verify, compare the `RepoDigests` values you get from `podman inspect quay.io/libpod/alpine` after freshly pulling the image, and after a `podman save`/`podman rmi`/`podman load` cycle.

**Comment by edsantiago** [comment]
Thank you. Looks like a reasonable analysis. How do we fix that?

**Comment by nalind** [comment]
Hmm, this has been happening for a while, since I see it happening with both podman 4.9.4 and 5.0.0 as well, so it might not be the only cause.
But to your question, buildah uses a "dir:" location for the cache to avoid changing the image while caching it, and a custom "copy" test helper since it doesn't want to depend on having skopeo available at test time.  But podman already expects skopeo to be present, so `skopeo copy` should be up to the task.

**Comment by nalind** [comment]
> AFAIK the libimage code treats a non empty platform as must pull or something like that if I remember correctly However I haven't verified that.

I had suspected as much, but at most, that should be changing the pull policy to "if-newer" [under the covers](https://github.com/containers/common/blob/1a97fd11a5f298f2626b2b1d6f06a3f483d00af1/libimage/pull.go#L559), and I'm pretty sure that's already the value that's being passed in.

**Comment by Luap99** [comment]
> > AFAIK the libimage code treats a non empty platform as must pull or something like that if I remember correctly However I haven't verified that.
> 
> I had suspected as much, but at most, that should be changing the pull policy to "if-newer" [under the covers](https://github.com/containers/common/blob/1a97fd11a5f298f2626b2b1d6f06a3f483d00af1/libimage/pull.go#L559), and I'm pretty sure that's already the value that's being passed in.

Yeah not sure, but the fact is main is passing but this PR is not so the problem must be in the diff here and thankfully the diff is small so this is my best guess as of why. I can try to reproduce tomorrow.

**Comment by edsantiago** [comment]
>What I do see however is https://github.com/containers/buildah/pull/5478
>
>if platform.OS == "" {
>	platform.OS = runtime.GOOS
>}

Nice catch, @Luap99 . Removing those three lines (and the `import runtime`) appears to fix the problem.

**Comment by rhatdan** [comment]
That was added to make `podman build --arch arm64` work

**Comment by Luap99** [comment]
> That was added to make `podman build --arch arm64` work

Sure but it fixed it in  broken way, the default pull policy is ifmissing and not ifnewer which the code uses now due the well other not so great behavior in libimage. 
And reading your PR your intention is to only set TARGETARCH not change the pull policy so this is definitely broken and must be fixed before the buildah release

**Comment by edsantiago** [comment]
Treadmill has been broken for a few weeks, vendoring one of the c-* modules:
```
CGO_ENABLED=1 GOOS=linux GOARCH=amd64 go build \
         \
        -ldflags '-X github.com/containers/podman/v5/libpod/define.gitCommit=71ed11ece1223fd847c759b9ce5af180af370b62 -X github.com/containers/podman/v5/libpod/define.buildInfo=1720019142 -X github.com/containers/podman/v5/libpod/config._installPrefix=/usr/local -X github.com/containers/podman/v5/libpod/config._etcDir=/etc -X github.com/containers/podman/v5/pkg/systemd/quadlet._binDir=/home/esm/bin/x86_64 -X github.com/containers/common/pkg/config.additionalHelperBinariesDir= ' \
        -tags "   systemd libsubid exclude_graphdriver_devicemapper seccomp " \
        -o bin/podman ./cmd/podman
# github.com/containers/podman/v5/libpod
libpod/reset.go:262:26: undefined: blobinfocache.CleanupDefaultCache
make: *** [Makefile:380: bin/podman] Error 1
```
cc @mheon

**Comment by nalind** [comment]
Buildah's main branch is currently using containers/image v5.31.1, and podman's main branch is currently using an untagged commit on the containers/image main branch which is newer than that.  Attempting to vendor in buildah is apparently pulling the version of image that buildah uses, which doesn't supply the new function that libpod already depends on.

**Comment by edsantiago** [comment]
https://github.com/containers/buildah/pull/5585 has just merged. Treadmill is now failing with:
```console
$ make
CGO_ENABLED=1 GOOS=linux GOARCH=amd64 go build \
         \
        -ldflags '-X github.com/containers/podman/v5/libpod/define.gitCommit=e7bf0e98bbaebe5dd703c2080cf03fecf244720c -X github.com/containers/podman/v5/libpod/define.buildInfo=1721044095 -X github.com/containers/podman/v5/libpod/config._installPrefix=/usr/local -X github.com/containers/podman/v5/libpod/config._etcDir=/etc -X github.com/containers/podman/v5/pkg/systemd/quadlet._binDir=/home/esm/bin/x86_64 -X github.com/containers/common/pkg/config.additionalHelperBinariesDir= ' \
        -tags "   systemd libsubid exclude_graphdriver_devicemapper seccomp " \
        -o bin/podman ./cmd/podman
test -z "-Z" || chcon -t container_runtime_exec_t bin/podman
CGO_ENABLED=1 GOOS=linux GOARCH=amd64 go build \
         \
        -ldflags '-X github.com/containers/podman/v5/libpod/define.gitCommit=e7bf0e98bbaebe5dd703c2080cf03fecf244720c -X github.com/containers/podman/v5/libpod/define.buildInfo=1721044108 -X github.com/containers/podman/v5/libpod/config._installPrefix=/usr/local -X github.com/containers/podman/v5/libpod/config._etcDir=/etc -X github.com/containers/podman/v5/pkg/systemd/quadlet._binDir=/home/esm/bin/x86_64 -X github.com/containers/common/pkg/config.additionalHelperBinariesDir= ' \
        -tags "remote exclude_graphdriver_btrfs btrfs_noversion exclude_graphdriver_devicemapper containers_image_openpgp" \
        -o bin/podman-remote ./cmd/podman
package github.com/containers/podman/v5/cmd/podman
        imports github.com/containers/podman/v5/cmd/podman/common
        imports github.com/containers/buildah/define
        imports github.com/containers/common/libimage: build constraints exclude all Go files in vendor/github.com/containers/common/libimage
make: *** [Makefile:393: bin/podman-remote] Error 1
buildah-vendor-treadmill-x: 'make' failed with new buildah. Cannot continue.
```

**Comment by rhatdan** [comment]
Working on it.

**Comment by rhatdan** [comment]
@nalind @edsantiago Looks like this is the culprit
[t.txt](https://github.com/user-attachments/files/16234661/t.txt)

**Comment by rhatdan** [comment]
@aaronlehmann  PTAL

**Comment by nalind** [comment]
One option in https://github.com/containers/buildah/pull/5628.

**Comment by nalind** [comment]
> One option in [containers/buildah#5628](https://github.com/containers/buildah/pull/5628).

Another option, which all things being equal, I prefer, in https://github.com/containers/common/pull/2089.

**Comment by edsantiago** [comment]
Aaaaand we're back to `undefined: blobinfocache.CleanupDefaultCache`. Dunno if it's the latest podman commit or the latest buildah one. Somehow c-image looks correct on buildah, but treadmill now wants to change it to `v5.31.1`

**Comment by nalind** [comment]
https://github.com/containers/buildah/pull/5627 was a downgrade.

**Comment by edsantiago** [comment]
Gah. I completely forgot about `remote`.

One possible solution will be to use a custom socket for each process, maybe derived from `$BATS_SUITE_TEST_NUMBER` or maybe pseudorandom. That is going to be trickier than anything I can do in one day.

Good news is, though, THIS EFFORT IS WORTH IT! (parallelizing). Wow, 15 minutes for bud tests.

**Comment by edsantiago** [comment]
\*sniffle\*. Look at this. Passing bud tests, all in less than sixteen minutes. (For comparison: bud tests are by far our biggest slowdown, taking ~35 minutes on a CI run).

My changes here are absolutely not ready for prime time, but I think they're in good enough shape that someone else should be able to carry them through without much fuss.

**Comment by Luap99** [comment]
I was counting on the parallel work to get this time down.

> My changes here are absolutely not ready for prime time, but I think they're in good enough shape that someone else should be able to carry them through without much fuss.

Just to be clear we need to get the parallel test going in buildah first? Once that works we can enable it here for bud tests?!
